python/hawkey/subject-py.c

```diff
@@ -272,41 +289,95 @@ reldep_possibilities_real(_SubjectObject *self, PyObject *args, PyObject *kwds)
         csack, flags);
     return possibilitiesToPyObject(iter, sack);
 }
+
 static PyObject *
-get_best_solution(_SubjectObject *self, PyObject *args, PyObject *kwds)
+get_best_parser(_SubjectObject *self, PyObject *args, PyObject *kwds, HyNevra *nevra)
 {
     PyObject *sack;
     DnfSack *csack;
-    PyObject *form = NULL;
-    PyObject *icase = NULL;
+    PyObject *forms = NULL;
     PyObject *with_nevra = NULL;
     PyObject *with_provides = NULL;
     PyObject *with_filenames = NULL;
-    const char *kwlist[] = {"sack", "form", "icase", "with_nevra", "with_provides",
-        "with_filenames", NULL};
+    const char *kwlist[] = {"sack", "with_nevra", "with_provides", "with_filenames", "forms", NULL};
 
-    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O!|OO!O!O!O!", (char**) kwlist, &sack_Type, &sack,
-        &form, &PyBool_Type, &icase, &PyBool_Type, &with_nevra, &PyBool_Type, &with_provides,
-        &PyBool_Type, &with_filenames)) {
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O!|O!O!O!O", (char**) kwlist, &sack_Type, &sack,
+        &PyBool_Type, &with_nevra, &PyBool_Type, &with_provides,
+        &PyBool_Type, &with_filenames, &forms)) {
         return NULL;
     }
     HyForm *cforms = NULL;
-    if (form != NULL) {
-        cforms = fill_form(form);
+    if ((forms != NULL) && (forms != Py_None) &&
+        ((!PyList_Check(forms)) || (PyList_Size(forms) > 0))) {
+        cforms = fill_form(forms);
         if (cforms == NULL)
             return NULL;
     }
-    gboolean c_icase = icase!=NULL && PyObject_IsTrue(icase);
-    gboolean c_with_nevra = icase == NULL || PyObject_IsTrue(with_nevra);
-    gboolean c_with_provides = icase == NULL || PyObject_IsTrue(with_provides);
-    gboolean c_with_filenames = icase == NULL || PyObject_IsTrue(with_filenames);
+    gboolean c_with_nevra = with_nevra == NULL || PyObject_IsTrue(with_nevra);
+    gboolean c_with_provides = with_provides == NULL || PyObject_IsTrue(with_provides);
+    gboolean c_with_filenames = with_filenames == NULL || PyObject_IsTrue(with_filenames);
     csack = sackFromPyObject(sack);
+    HyQuery query = hy_subject_get_best_solution(self->pattern, csack, cforms, nevra, self->icase,
+                                                 c_with_nevra, c_with_provides, c_with_filenames);
+
+    return queryToPyObject(query, sack, &query_Type);
+}
+
+static PyObject *
+get_best_query(_SubjectObject *self, PyObject *args, PyObject *kwds)
+{
     HyNevra nevra = NULL;
 
-    HyQuery query = hy_subject_get_best_solution(self->pattern, csack, cforms, &nevra, c_icase,
-                                                 c_with_nevra, c_with_provides, c_with_filenames);
+    PyObject *py_query = get_best_parser(self, args, kwds, &nevra);
+    if (nevra != NULL) {
+        g_clear_pointer(&nevra, hy_nevra_free);
+    }
+    Py_INCREF(py_query);
```

Comment:
get_best_parser() may return NULL. But Py_INCREF() must not be used with NULL argument.
Use Py_XINCREF().